File: Hub/src/main/java/com/salcedo/rapbot/motor/MotorActor.java
package com.salcedo.rapbot.motor;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import com.salcedo.rapbot.snapshot.RegisterSubSystemMessage;
import com.salcedo.rapbot.snapshot.SnapshotMessage;
import com.salcedo.rapbot.snapshot.TakeSnapshotMessage;

public final class MotorActor extends AbstractActor {
    private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);
    private final MotorService motorService;

    public MotorActor(final MotorService motorService) {
        this.motorService = motorService;
    }

    public static Props props(final MotorService motorService) {
        return Props.create(MotorActor.class, motorService);
    }

    @Override
    public void preStart() throws Exception {
        release();
        context().system().eventStream().publish(new RegisterSubSystemMessage(self()));
        context().system().eventStream().subscribe(self(), TakeSnapshotMessage.class);
    }

    @Override
    public void postStop() throws Exception {
        release();
    }

    private void release() {
        motorService.release();
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(MotorRequest.class, this::drive)
                .match(TakeSnapshotMessage.class, this::snapshot)
                .build();
    }

    private void snapshot(TakeSnapshotMessage message) {
        final ActorRef sender = sender();
        motorService.state()
                .thenAccept(response -> sender.tell(new SnapshotMessage(message.getUuid(), response), self()));
    }

    private void drive(final MotorRequest request) {
        final ActorRef sender = sender();
        motorService.drive(request).thenAccept(response -> self().tell(response, sender));
    }
}

File: Hub/src/main/scala/com/salcedo/rapbot/motor/MotorActor.scala
package com.salcedo.rapbot.motor

import java.util.UUID

import akka.actor.Status.Success
import akka.actor.{Actor, ActorLogging, Props}
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.stream.scaladsl.{JsonFraming, Sink}
import akka.stream.{ActorMaterializer, ActorMaterializerSettings}
import com.salcedo.rapbot.motor.MotorActor._
import com.salcedo.rapbot.remote.ActorBreaker
import com.salcedo.rapbot.serialization.JSON.{read, write}
import com.salcedo.rapbot.snapshot.SnapshotActor.TakeSubSystemSnapshot

import scala.concurrent.Future

object MotorActor {

  sealed case class Command(value: Int)

  object Forward extends Command(1)

  object Backward extends Command(2)

  object Brake extends Command(3)

  object Release extends Command(4)

  case class Motor(speed: Int, command: Command)

  case class Vehicle(backLeft: Motor, backRight: Motor)

  case class Pushed(version: UUID)

  def props(uri: Uri): Props = Props(new MotorActor(uri))
}

class MotorActor(val uri: Uri) extends Actor with ActorBreaker with ActorLogging {
  implicit val materializer: ActorMaterializer = ActorMaterializer(ActorMaterializerSettings(context.system))

  val http = Http(context.system)
  var vehicle = Vehicle(Motor(0, Release), Motor(0, Release))
  var version: UUID = nextVersion
  var remoteCall: Future[UUID] = Future.successful(version)

  override def preStart(): Unit = {
    context.system.eventStream.subscribe(self, classOf[Vehicle])
    send()
  }

  override def receive: PartialFunction[Any, Unit] = {
    case vehicle: Vehicle => this.update(vehicle)
    case Pushed(v) => this.push(v)
    case _: TakeSubSystemSnapshot => this.snapshot()
  }

  def update(vehicle: Vehicle): Unit = {
    if (this.vehicle.equals(vehicle)) return

    this.version = nextVersion
    this.vehicle = vehicle

    send()
  }

  private def nextVersion = {
    UUID.randomUUID()
  }

  def push(pushedVersion: UUID): Unit = {
    if (version.equals(pushedVersion)) {
      log.debug("Received a pushed message for '{}' version", version)
      return
    }

    send()
  }

  def send(): Unit = {
    if (!remoteCall.isCompleted) return

    val request = HttpRequest(
      uri = uri.withPath(Uri.Path("/motors")),
      method = HttpMethods.PUT,
      entity = HttpEntity(ContentTypes.`application/json`, write(vehicle))
    )

    val future = http.singleRequest(request).flatMap {
      case HttpResponse(StatusCodes.OK, _, entity, _) =>
        parseVehicle(entity)
      case HttpResponse(statusCode, _, entity, _) =>
        entity.discardBytes()
        sys.error(s"Received a response with a status code other than 200 OK. Status: $statusCode")
    }(context.dispatcher)

    future.onComplete(_ => self ! Pushed(version))(context.dispatcher)
  }

  def snapshot(): Unit = {
    sender() ! Success(vehicle)
  }

  private def parseVehicle(entity: ResponseEntity): Future[Vehicle] = {
    entity.transformDataBytes(JsonFraming.objectScanner(256))
      .dataBytes
      .map(_.utf8String)
      .map(read(_, classOf[Vehicle]))
      .runWith(Sink.seq)
      .map(_.head)(context.dispatcher)
  }
}
